packages/react/src/hooks/useRef.ts

```diff
@@ -1,15 +1,20 @@
-import React, {useRef} from 'react'
+import {useRef, type RefObject, type ForwardedRef} from 'react'
 
 /**
  * There are some situations where we only want to create a new ref if one is not provided to a component
  * or hook as a prop. However, due to the `rules-of-hooks`, we cannot conditionally make a call to `React.useRef`
  * only in the situations where the ref is not provided as a prop.
  * This hook aims to encapsulate that logic, so the consumer doesn't need to be concerned with violating `rules-of-hooks`.
  * @param providedRef The ref to use - if undefined, will use the ref from a call to React.useRef
- * @type TRef The type of the RefObject which should be created.
+ * @type T The type of the RefObject which should be created.
  */
-export function useProvidedRefOrCreate<TRef>(providedRef?: React.RefObject<TRef>): React.RefObject<TRef> {
-  const createdRef = useRef<TRef>(null)
 
-  return providedRef ?? createdRef
+export function useProvidedRefOrCreate<T>(providedRef?: null | RefObject<T> | ForwardedRef<T>): RefObject<T> {
+  const createdRef = useRef<T>(null)
+
+  if (providedRef) {
+    return typeof providedRef === 'function' ? createdRef : providedRef
+  }
```

Comment:
Did you run into type issues with this hook, prompting this change? I don't think we have test coverage on this file, but we use it in a few places. Are we confident this won't cause regressions as the logic here is now different.

Reply:
Yea I did run into some type issues, and it looks like we've run into the same ones elsewhere too. In all of the other places that we use `useProvidedRefOrCreate` we override the types with something like

```tsx
useProvidedRefOrCreate(ref as RefObject<HTMLDivElement>)
```

which suggests to me that the types of `useProvidedRefOrCreate` aren't right to start with, which is why I wanted to improve the types.

It's a good point that I shouldn't change the implementation though, and to be honest I probably shouldn't touch it in this PR at all.

I'll dig a bit deeper into this and raise a follow-up PR to address this

Reply:
Reverted in `8102bdee4e858f8c34ef957d0d677e023601cbd8`